Adopt more smart pointers in IntersectionObserver

rwlbuis opened the pull request:
#### e864738237198a14b63c6f519be1023b951f00fe
<pre>
Adopt more smart pointers in IntersectionObserver
<a href="https://bugs.webkit.org/show_bug.cgi?id=273478">https://bugs.webkit.org/show_bug.cgi?id=273478</a>

Reviewed by Chris Dumez.

Adopt more smart pointers in IntersectionObserver based on the
[alpha.webkit.UncountedLocalVarsChecker] warning.

* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::IntersectionObserver::observe):
(WebCore::IntersectionObserver::unobserve):
(WebCore::IntersectionObserver::disconnect):
(WebCore::IntersectionObserver::targetDestroyed):
(WebCore::computeClippedRectInRootContentsSpace):
(WebCore::IntersectionObserver::nowTimestamp const):
(WebCore::IntersectionObserver::notify):
(WebCore::IntersectionObserver::isReachableFromOpaqueRoots const):

Canonical link: <a href="https://commits.webkit.org/278315@main">https://commits.webkit.org/278315@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c23d94bd4550b65cf750f9bb3111d5bd461d533a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49853 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29140 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2104 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53096 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/530 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52154 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/97 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40675 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51951 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26700 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42890 "Found 1 new API test failure: TestWebKitAPI.HSTS.Preconnect (failure)") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21800 "Found 1 new API test failure: /WPE/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24133 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/89 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; 13 flakes 9 failures; Uploaded test results; 9 flakes 5 failures; Running compile-webkit-without-change") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8223 "Built successfully") | 
| | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46153 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; 6 api tests failed or timed out; Running re-run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/129 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54677 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/95 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48062 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26204 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43058 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47094 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11004 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27065 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25932 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/IntersectionObserver.cpp:
```diff
@@ -579,7 +583,8 @@ void IntersectionObserver::notify()
 {
     for (auto& target : m_observationTargets) {
-        if (auto* element = target.get(); containsWebCoreOpaqueRoot(visitor, element))
+        SUPPRESS_UNCOUNTED_LOCAL auto* element = target.get();
```
@rniwa @geoffreygaren Do we really want to start spreading the `SUPPRESS_UNCOUNTED_LOCAL` use for things that are false positives in the static analyzer that we should be able to address in the near future? Seems unfortunate to make the codebase look worse just because of a static analyzer bug.

rniwa replied:
So this one, I'm not sure if we can call it a bug in the static analyzer. `containsWebCoreOpaqueRoot` calls `Visitor::containsOpaqueRoot`, which in turn calls `didFindOpaqueRoot`, which is a non-inline function. There isn't really a way that static analyzer can detect this is a trivial function unless we hard-code the function name in the static analyzer or something.

geoffreygaren replied:
I'd prefer to explicitly annotate containsWebCoreOpaqueRoot() as trivial / non-deleting. Otherwise we'll have to suppress in hundreds of places. I guess it would require a small feature in the static analyzer to offer that annotation.

rniwa commented on Source/WebCore/page/IntersectionObserver.cpp:
```diff
@@ -143,9 +143,13 @@ IntersectionObserver::IntersectionObserver(Document& document, Ref<IntersectionO
+    } else {
+        SUPPRESS_UNCOUNTED_LOCAL auto* frame = document.frame();
+        if (frame) {
+            SUPPRESS_UNCOUNTED_LOCAL auto* localFrame = dynamicDowncast<LocalFrame>(frame->mainFrame());
```
These two look like a bug in static analyzer. Everything in this code block is trivial so we shouldn't be needing to suppress the warning.

rwlbuis replied:
I removed this part.

rniwa commented on Source/WebCore/page/IntersectionObserver.cpp:
```diff
@@ -526,12 +530,12 @@ std::optional<ReducedResolutionSeconds> IntersectionObserver::nowTimestamp() con
         return std::nullopt;
 
-    auto* context = m_callback->scriptExecutionContext();
+    SUPPRESS_UNCOUNTED_LOCAL auto* context = m_callback->scriptExecutionContext();
```
This one is warned because `frozenNowTimestamp` down below is a non-inlined function call. Even though we don't use context after that point, static analyzer isn't smart enough to see that. I'm not sure if it's possible to "fix" this in the static analyzer.

rniwa replied:
Okay, so I think our suggestion here is to do the following:
```cpp
    RefPtr<DOMWindow> window;
    {
        auto* context = m_callback->scriptExecutionContext();
        if (!context)
             return std::nullopt;
        auto& document = downcast<Document>(*context);
        window = document.domWindow();
        if (!window)
            return std::nullopt;
    }
    return window->frozenNowTimestamp();
```

rwlbuis replied:
Done.

rwlbuis committed:
Adopt more smart pointers in IntersectionObserver
https://bugs.webkit.org/show_bug.cgi?id=273478

Reviewed by Chris Dumez.

Adopt more smart pointers in IntersectionObserver based on the
[alpha.webkit.UncountedLocalVarsChecker] warning.

* Source/WebCore/page/IntersectionObserver.cpp:
(WebCore::IntersectionObserver::observe):
(WebCore::IntersectionObserver::unobserve):
(WebCore::IntersectionObserver::disconnect):
(WebCore::IntersectionObserver::targetDestroyed):
(WebCore::computeClippedRectInRootContentsSpace):
(WebCore::IntersectionObserver::nowTimestamp const):
(WebCore::IntersectionObserver::notify):
(WebCore::IntersectionObserver::isReachableFromOpaqueRoots const):

Canonical link: https://commits.webkit.org/278315@main
